Enabled ttxt as a lib

iulianbarbu opened the pull request:
Made the crate as a lib (in addition to being a binary) but..

Did a bunch on unnecessary changes too:
- fixed clippy errors showing up in my editor (curious if you get any on your end, I am using 1.82)
- enabled fmt to work on my IDE (nightly-2024-04-10)
- upgraded packages (my crates.nvim extension showed me a few warnings with newer packages)
- enabled taplo by adding a config file resembling the one in `polkadot-sdk`
- one of the tests fails on my machine (oh_god2), so I ignored it.
- upgrading the packages required at some point to drop the `account_id()` call on keyring, and use `public_key()` instead. I think this should be alright but lmk if you have another opinion.

I can revert the unnecessary changes if you want or find it important for some reason.

Closes #18

michalkucharczyk commented on src/transaction.rs:
```diff
@@ -133,3 +134,86 @@ pub trait TransactionsSink<H: BlockHash>: Send + Sync {
+
+#[derive(Subcommand)]
+/// Send transactions to the node using different scenarios.
+pub enum SendingScenario {
```
maybe this could be moved to the dedicated file?

iulianbarbu replied:
I moved this back to the cli module. I created a new module called `scenario`, where I developed some logic related to the idea of a `scenario`. I kept the `SendingScenario` in the `cli` though (seeing it as a subcommand made me think that's its place, but lmk if you have another idea).

iulianbarbu replied:
>  I kept the SendingScenario in the cli though

Spoke to soon: [cf898d4](https://github.com/michalkucharczyk/tx-test-tool/pull/22/commits/cf898d4a74adc76c20bdcf86d5b428657fd7ea1f)

iulianbarbu commented on src/transaction.rs:
```diff
@@ -44,7 +45,7 @@ pub trait TransactionStatusIsTerminal {
 pub enum TransactionStatus<H> {
 	Validated,
-	Broadcasted(u32),
+	Broadcasted,
```
Because we implement a `From` trait for `TransactionStatus`, from `subxt::tx::TxStatus`, which seems to have been updated to not contain the peers num (subxt 0.37.0 -> 0.39.0). Not super sure how useful the num_peers was, and not sure either why `TxStatus` has changed in between. Have to look closer.

michalkucharczyk replied:
Should be fine.

michalkucharczyk commented on src/main.rs:
```diff
@@ -353,71 +38,49 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
+					.with_scenario_type(scenario.clone())
+					.with_block_monitoring()
+					.with_chain_type(chain.clone())
+					.with_send_threshold(*send_threshold as usize);
```
We need to handle `remark` param too. 
So `.with_transfer_tx_recipe()` or `with_remark_tx_recipe` shall be called.
Also in `sub` branch.

michalkucharczyk replied:
oh, it is handled below - but transfer needs to be default.

iulianbarbu replied:
Indeed. Added it for both sub & eth cases. Change here: [2319bdf](https://github.com/michalkucharczyk/tx-test-tool/pull/22/commits/2319bdf4509881f8c4c8bf186444a2b605fdb8d5)

michalkucharczyk commented on src/scenario.rs:
```diff
@@ -0,0 +1,462 @@
+	}
+
+	pub fn with_block_monitoring(mut self) -> Self {
+		self.does_block_monitoring = true;
```
```suggestion
		self.does_block_monitoring = false;
```

iulianbarbu replied:
I realized I didn't take into account the `--block-monitoring` flag in `main.rs`. So this change isn't needed, we just need to set the block monitoring only if the flag is set via CLI, while in API, we just need to use the setter whenever we want to use block monitoring. Change here: [2319bdf](https://github.com/michalkucharczyk/tx-test-tool/pull/22/commits/2319bdf4509881f8c4c8bf186444a2b605fdb8d5)

michalkucharczyk commented on src/scenario.rs:
```diff
@@ -0,0 +1,462 @@
+	}
+
+	pub fn with_transfer_tx_recipe(mut self) -> Self {
+		self.tx_recipe = Some(TransactionRecipe::transfer());
```
This should be default.

iulianbarbu replied:
Yes, added it by default when constructing the scenario, and can be changed if `remark` is set. Change here: [2319bdf](https://github.com/michalkucharczyk/tx-test-tool/pull/22/commits/2319bdf4509881f8c4c8bf186444a2b605fdb8d5)

michalkucharczyk commented on src/scenario.rs:
```diff
@@ -0,0 +1,462 @@
+
+impl ScenarioBuilder {
+	pub fn with_scenario_type(mut self, scenario_type: ScenarioType) -> Self {
+		self.scenario_type = Some(scenario_type);
```
hm, not sure if `ScenarioType` type should be exposed.

iulianbarbu replied:
Ah, I remember your former suggestion. The builder here might take the form of `with_start_account_id`, `with_last_account_id`, `with_nonce_from`, `with_txs_count`. Based on a combination of these parameters we can derive if the scenario type is `OneShot` (e.g. only `start_id` and `nonce_from`  is given) , `FromSingleAccount` (only `start_id`, `nonce_from` and `txs_count` are given) or `FromManyAccounts` (we give all the params). WDYT?

michalkucharczyk replied:
Your idea makes sense, we can guess scenario type.

michalkucharczyk replied:
Just realized we cannot get rid of this type completely - because of CLI/main.

iulianbarbu replied:
Indeed, still kept the builder API that's not using the `ScenarioType`. The refactoring is here: [9f14440](https://github.com/michalkucharczyk/tx-test-tool/pull/22/commits/9f14440a0747fa8f44b382607d6bcc36649465e7). Resolving this but if you have comments lets start a new thread.

michalkucharczyk commented on src/main.rs:
```diff
@@ -353,71 +38,49 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
+				let mut scenario_builder = ScenarioBuilder::default()
+					.with_rpc_uri(ws.to_string())
+					.with_scenario_type(scenario.clone())
+					.with_block_monitoring()
```
This should take bool: 
```
.with_block_monitoring(block_monitoring)
```

iulianbarbu replied:
Hmm, I actually imagined it as we either call the setter (setting it to `true`), or we don't (initialized with false and kept that way). I was thinking we're not interested in changing it once we set it to `true`, so no need to give the option to "unset it". WDYT?

iulianbarbu replied:
Addressed your initial request here: [9f14440](https://github.com/michalkucharczyk/tx-test-tool/pull/22/commits/9f14440a0747fa8f44b382607d6bcc36649465e7).

michalkucharczyk commented on src/main.rs:
```diff
@@ -353,71 +38,57 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
+					.with_send_threshold(*send_threshold as usize);
+
+				if block_monitor {
+					scenario_builder = scenario_builder.with_block_monitoring();
```
I think
```
.with_block_monitoring(*block_monitor);
```
would be nicer.

iulianbarbu replied:
no strong opinion on this, will change

michalkucharczyk commented on src/scenario.rs:
```diff
@@ -0,0 +1,479 @@
+	/// - `does_block_monitoring` is set to `false`.
+	/// - `installs_ctrl_c_stop_hook` is set to `false`.
+	/// - `send_threshold` is set to `1000`.
+	pub fn init() -> Self {
```
```suggestion
	pub fn new() -> Self {
```

iulianbarbu replied:
Done here: [9f14440](https://github.com/michalkucharczyk/tx-test-tool/pull/22/commits/9f14440a0747fa8f44b382607d6bcc36649465e7)

michalkucharczyk commented on src/scenario.rs:
```diff
@@ -0,0 +1,479 @@
+}
+
+impl ScenarioExecutor {
+	pub async fn execute_txs<T: TxTask + 'static>(
```
maybe just execute?
```suggestion
	pub async fn execute<T: TxTask + 'static>(
```

iulianbarbu replied:
Done here: [ae56f19](https://github.com/michalkucharczyk/tx-test-tool/pull/22/commits/ae56f19ba72e27a8bd886dd8b288351a0a5d7666)

michalkucharczyk commented on src/lib.rs:
```diff
@@ -0,0 +1,102 @@
+pub mod scenario;
+pub mod subxt_api_connector;
+pub mod subxt_transaction;
+pub mod transaction;
```
One more clean-up that could be done in this PR is exposing only what is required in zombienet / main.

iulianbarbu replied:
Zombienet tests will need the `ScenarioBuilder`, `ScenarioExecutor`, the `DefaultTxTask` and the transaction types: `EthTransaction` & `SubstrateTransaction`.

`main.rs` needs a few more.

There are a few more types that can be limited to `pub(crate)` or maybe remove `pub(crate)` altogether, but if done so Rust compiler complains that these types are related to other types that have higher visibility (meaning `pub`). In practice we can lower the visibility for all the types that are not directly used in `main.rs` & `zombienet` tests, and suppress the associated visibility warnings (which are a lot), but I decided to expose the types that rust compiler complains about, even if not directly used. I wouldn't overindex on the pub API that much, considering the way zombienet tests will be developed will rely on the types I mentioned. The rest of the types can extend `ttxt` for library users (e.g. add new tx builder, sink & txs), but at this time we're the only users.

I wouldn't publish `ttxt` on `crates.io` yet, but use it based on a git dependency in our tests for a bit of time until we decide more on how the API should look like (and after raising a bit more the tests coverage). WDYT?

iulianbarbu commented:
In latest batch of commits I added:
 - hash fix you gave me on Element
 - tip cli flag patch you sent me on Element
 - added `derive_accounts` test in `subxt_transaction`
- lib usage example

michalkucharczyk commented on src/main.rs:
```diff
@@ -353,71 +55,47 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
+					.with_send_threshold(*send_threshold as usize)
+					.with_block_monitoring(*block_monitor)
+					.with_watched_txs(!unwatched)
+					.with_tip(*tip);
```
tip could be owevritten by lines 74-75 (remark recipe setting).
Maybe tip should be a dedicated field in `ScenarioBuilder` - otherwise it could be overwritten when recipe is set after. (or should be moved from old recipe when new is set).

iulianbarbu replied:
Good catch! Made `with_tip` update its own field and the recipe, while setting the recipe uses the previously set `tip` (0 if not set, and the recipe will be updated once a new tip is set).

michalkucharczyk commented on src/main.rs:
```diff
@@ -353,71 +55,47 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
+					.with_send_threshold(*send_threshold as usize)
+					.with_block_monitoring(*block_monitor)
+					.with_watched_txs(!unwatched)
+					.with_tip(*tip);
```
same here

iulianbarbu replied:
same as before

michalkucharczyk commented on src/scenario.rs:
```diff
@@ -0,0 +1,548 @@
+	/// [`subxt_transaction::derive_accounts`]), used to generate signer accounts, which are assumed
+	/// to be initialised as development accounts with balances by the networks where the
+	/// transactions will be executed.
+	pub fn with_start_id(mut self, start_id: String) -> Self {
```
Could we easily have `u32` param here?
Maybe we could add new function `with_account_id` for passing strings?

 `with_start_id("0").with_last_id(99)` is not intuitive.

iulianbarbu replied:
Yes, I agree it is not intuitive. I added `with_account_id` now, and the API looks cleaner.

michalkucharczyk commented on src/execution_log.rs:
```diff
@@ -179,6 +183,7 @@ pub trait ExecutionLog: Sync + Send {
 
 #[derive(Debug)]
+/// A default implementation of an execution log.
 pub struct DefaultExecutionLog<H: BlockHash> {
```
As this is now public, I would rename it, `DefaultExecutionLog` is not a most fortunate name.
I'd propose `TransactionExecutionLog`, but I am open to other ideas.

iulianbarbu replied:
Done.

michalkucharczyk commented on src/scenario.rs:
```diff
@@ -0,0 +1,586 @@
+	/// The builder is already initialised with a transfer transaction recipe for the built
+	/// transactions, and this API
+	pub fn with_transfer_tx_recipe(mut self) -> Self {
+		let mut recipe = TransactionRecipe::transfer();
+		recipe.tip = self.tip;
+		self.tx_recipe = Some(recipe);
```
That would be nicer:
```suggestion
		self.tx_recipe = Some(TransactionRecipe::transfer(self.tip));
```
(or `transfer_with_tip` ?)

iulianbarbu replied:
Yeh, this is better. I think we don't need to store the `tip` on the builder (and have a dedicated setter for it), but instead initialize the builder with a transfer recipe with tip 0, and then if people want to set a certain tip, they would need to set the recipe as: `with_transfer_recipe(mut self, tip: u128)`, or `with_remark_recipe(mut self, tip: u128)`. WDYT?

michalkucharczyk replied:
I would keep it in builder. It is a way to control priority of the transaction. There will be one more - mortality - which will also go to the recipe. I feel that having dedicated setters will be easier on both sides - user and builder implementation. Also more recipes may come - eg nft minting or similar.

iulianbarbu replied:
At that time probably a dedicated recipe builder sounds better, but yeh, nesting the recipe building in the scenario is also an option for the time being. Will change it quickly

iulianbarbu replied:
Done here: [b1d48c2](https://github.com/michalkucharczyk/tx-test-tool/pull/22/commits/b1d48c212749f0ef791b963f4e8a90340ce6dfba).
